lib/history/sqlite.go

```diff
@@ -0,0 +1,379 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package history
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+
+	"github.com/gravitational/trace"
+	"github.com/jmoiron/sqlx"
+	_ "github.com/mattn/go-sqlite3" // initialize sqlite3
+	log "github.com/sirupsen/logrus"
+)
+
+// SQLiteTimeline represents a timeline of cluster status events. The timeline
+// can hold a specified amount of events and uses a FIFO eviction policy.
+// Timeline events are stored in a local sqlite database.
+//
+// Implements Timeline
+type SQLiteTimeline struct {
+	// clock is used to record event timestamps.
+	clock clockwork.Clock
+	// capacity specifies the max number of events that can be stored in the timeline.
+	capacity int
+	// size specifies the current number of events stored in the timeline.
+	size int
+	// database points to underlying sqlite database.
+	database *sqlx.DB
+	// lastStatus holds the last recorded cluster status.
+	lastStatus *pb.SystemStatus
+	// mu locks timeline access.
+	mu sync.Mutex
+}
+
+// SQLiteTimelineConfig defines SQLiteTimeline configuration.
+type SQLiteTimelineConfig struct {
+	// DBPath specifies the database location.
+	DBPath string
+	// Capacity specifies the max number of events that can be stored in the timeline.
+	Capacity int
+	// Clock will be used to record event timestamps.
+	Clock clockwork.Clock
+}
+
+// NewSQLiteTimeline initializes and returns a new SQLiteTimeline with the
+// specified configuration.
+func NewSQLiteTimeline(config SQLiteTimelineConfig) (*SQLiteTimeline, error) {
+	database, err := initSQLite(config.DBPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &SQLiteTimeline{
+		clock:    config.Clock,
+		capacity: config.Capacity,
+		size:     0,
+		database: database,
+		// TODO: store and recover lastStatus in case satellite agent restarts.
+		lastStatus: nil,
+	}, nil
+}
+
+// initSQLite initializes connection to database provided by dbPath and
+// initializes `events` table.
+func initSQLite(dbPath string) (*sqlx.DB, error) {
+	// TODO: Store sql db connection timeout as const
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	database, err := sqlx.ConnectContext(ctx, "sqlite3", dbPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if _, err := database.ExecContext(ctx, createTableEvents); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return database, nil
+}
+
+// RecordStatus records the differences between the previously stored status and
+// the provided status.
+func (t *SQLiteTimeline) RecordStatus(ctx context.Context, status *pb.SystemStatus) (err error) {
+	events := diffCluster(t.clock, t.lastStatus, status)
+	if len(events) == 0 {
+		return nil
+	}
+
+	t.mu.Lock()
```

Comment:
As a general rule, it is not advised to place any I/O operations under lock. Certainly not the database calls. What do you need it here for?